tensorrt_llm/_torch/pyexecutor/llm_request.py

```diff
@@ -75,21 +78,36 @@ def append(self, logits: torch.Tensor):
 
         assert logits.size(1) == self.beam_width, "Beam width mismatch"
 
-        new_position = logits.size(0) + self.position
+        position = 0 if not self._logits_indices else self._logits_indices[-1][1]
+        new_position = logits.size(0) + position
         if new_position > self.seq_length:
             raise ValueError(
                 f"LogitsStorage overflow. This storage can only hold {self.seq_length} logits "
-                f"({self.position} already filled) but trying to append {logits.size(0)} more logits"
+                f"({position} already filled) but trying to append {logits.size(0)} more logits"
             )
 
-        self._storage[self.position:new_position].copy_(logits,
-                                                        non_blocking=True)
-        self.last_position, self.position = self.position, new_position
+        self._storage[position:new_position].copy_(logits, non_blocking=True)
+        self._logits_indices.append((position, new_position))
+
+    def get_at(self, index: int) -> torch.Tensor | None:
+        if self._should_exclude_last:
+            index -= 1
+        try:
+            start, end = self._logits_indices[index]
+            return self._storage[start:end]
+        except IndexError:
+            return None
 
-    def get(self, all_logits=False):
-        start = 0 if all_logits else self.last_position
-        return self._storage[start:self.
-                             position] if self._storage is not None else None
+    def get_all(self):
+        try:
+            last_index = -2 if self._should_exclude_last else -1
```

Comment:
I assume there is a special case when the length of the container is 1, please cover that here.